### ISSUE TITLE
Enhancement: [no-floating-promises] add an 'allowForKnownSafePromises' option

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#7008]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises

### Description

For years, various community members have asked to add a type allowlist/blocklist or other _"stop complaining about these specific areas"_ for `no-floating-promises`:

* #2640
* #3192 _(ish)_
* #4722
* #4791
* #5231
* #5844
* #6017
* #7006

<sub>_(including clearly duplicate issues to show how this is a relatively more requested option, and to surface the different use cases requested for)_</sub>

Out of those, #5844's `node:test` is most convincing - as `node:test` is stable as of Node 20.0.0 (https://github.com/nodejs/node/issues/46642 -> https://github.com/nodejs/node/pull/46983), and returns `Promise<void>` for tests: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8730660c37a33dc0768b3f1e36b679294acc4988/types/node/test.d.ts#L134.

I propose we adopt the format discussed in #6017 & implemented in #4436 ([see v6 docs](https://v6--typescript-eslint.netlify.app/rules/prefer-readonly-parameter-types/#allow)) as a new `allow` option for `no-floating-promises`. We'd want to add some very explicitly, strongly worded descriptions in the docs as to why it's not an option we recommend using lightly.

### Fail

```typescript
async function test(name: string, executor: () => Promise<void>) {
  // ...
}

test("example", () => {});
```

### Pass

```typescript
import test from 'node:test';

test("example", () => {});
```

### Additional Info

cc @armano2 @bradzacher as past people who, like me, have previously disagreed with the suggestion of adding this option.

Edit: to be clear, this is roughly what users would be able to specify in their ESLint config _(pending naming bikeshedding)_:

```js
"@typescript-eslint/no-floating-promises": ["error", {
  "allowForKnownSafePromises": [
    { "from": "package", "name": "SafePromise", "package": "some-fancy-package"
  ]
}]
```

**Comment by bradzacher** [comment]
The `node:test` case is interesting and entirely unique case because as per the docs - it's actually inconsistent as to whether or not you should `await`:
- A top-level `test` call shouldn't be awaited (but it could be... The promise will resolve when the test finishes).
- A `test` call within a `describe` shouldn't be awaited (but it could be - the promise resolves immediately for this specific case).
- A `test` call within a `test` call should be awaited to ensure correct sequencing of tests.
    - worth noting thst again this is technically conditional! You should only await the sub-`test` if the callback is `async` or if the callback uses the `done` callback - otherwise the `await` is unnecessary.

This seems like horrible API design personally because deciding whether or not you should `await` the `test` call needs a flow chart - it's really complicated.

So in this one case it does make sense to have an option to special case around this standard API design.
I'd hope that someone would write a separate lint rule to enforce that `test` calls within a `test` function are correctly awaited.

----

As for a generic "ignore" option - personally I think there are very, very, very, very few cases where you want to always ignore the floating promise.

IMO If something always returns a promise any you literally never want to await it, then a better approach would be to patch the types to fix the return type and remove the promise so that nobody is ever tempted to `await` it accidentally.

If there's something you want to sometimes float and sometimes not then a disable comment or the `void` operator is a better approach so that you don't accidentally forget the cases where you don't want it floating and so that someone doesn't accidentally unfloat the intentionally floating case.

Examples like `node:test` are really, really rare in the ecosystem because people realise that they're not a good API design as they're confusing for end users.

TL;DR summary - I'm still against an ignore option, but I'm okay with special casing `node:test` in our rule.

**Comment by kaushalyap** [comment]
@JoshuaKGoldberg Any progress in this, I want `allow` option only run this rule on playwright tests files (`*.spec.ts`), enabling this rule is recommended by [Playwright docs](https://playwright.dev/docs/best-practices#lint-your-tests). In our case in front-end React project following statements nagging when rule is set to error

```
// next router navigation
router.push('/)

// Apollo mutations
createUser({...})
```

while could lower rule level to warning but it does not fail the CI when Playwright tests statement not containing `await`.

**Comment by bradzacher** [comment]
> I want allow option only run this rule on playwright tests files (*.spec.ts)

If you only want to enable the rule on playwright files you can do this with [eslint overrides](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-based-on-glob-patterns).

**Comment by kaushalyap** [comment]
@bradzacher Thanks, following worked

```
  "overrides": [
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@typescript-eslint/no-floating-promises": "error"
      }
    },
   ...
  ]
```

**Comment by phryneas** [comment]
@JoshuaKGoldberg suggested on Twitter that we could add an exception for certain branded types.

Context: using `createAsyncThunk` in Redux Toolkit or `endpoint.initiate()` in RTK Query will leave you with a promise that per definition will never throw.

```js
const myThunk = createAsyncThunk(...)

const promise = dispatch(myThunk())
// ^? will be a Promise for the action last dispatched by `myThunk`. This will either be a "fulfilled" action, or a "rejected" action, but either case it will always resolve and never reject.
// sometimes users want to await them, but oftentimes they want to just fire and forget this for the side effect and handling in the Redux store

const rejectingPromise = promise.unwrap()
// ^? this will be a promise that resolves in the case of a "fulfilled" action and rejects in the case of a "rejected" action
// essentially this is an opt-in into "I want to do try/catch".
```

This means that promise handling in Redux Toolkit is an "opt in" behaviour.

You *can* await the promise returned by `dispatch(myThunk())`, and you should do so if you want to do other calls after that promise has finished, but in 80-90% of cases you are probably not interested in the return value of `dispatch(myThunk())` at all:  
Handling of the resolved value happens in your Redux store, and handling of the rejected value also happens in your Redux store. 
These kind of promises are oftentimes "fire, forget & discard" and essentially an opt-in mechanism into asynchronous handling at call site - but most of the time there's no requirement to do so.

So, it would be nice if we could mark those promises with some kind of branded type to them to signal to eslint that these promises *can* be handled, but don't *need* to.

**Comment by JoshuaKGoldberg** [comment]
We talked about it a bit internally. The general consensus is that we think `allow` would be useful here _if_ library authors do a good job of branding Promises as you've described @phryneas. So, great! Yay! Filed https://github.com/reduxjs/redux-toolkit/issues/4101 to confirm on your end.

**Comment by JoshuaKGoldberg** [comment]
Marked as accepting PRs per https://github.com/reduxjs/redux-toolkit/issues/4101. Added to the v7 milestone to make sure we ship it on or before v7.0.0.

Note that out of the _dozens_ of times we have been asked for this option, that and `node:test` are the _only_ times it's made sense - and the `node:test` one is something we're hoping to fix on the Node side per https://github.com/nodejs/node/issues/51292. The docs for this option will need some _very strong wording_ expressing how _wrong_ it is to use this option _the vast majority of the time_. At the very least that means including very stern explanations for common reasons people _think_ they want this but are actually wrong.

In fact, I'd be in favor of naming this something intentionally obtuse and more specific than `allow`. Perhaps, `allowForKnownSafePromises`? Something even more obnoxious?

**Comment by mcollina** [comment]
I would want to add that this rule also applies to `Thenable`, that returns a Promise only when `.then()` is called. This should likely not apply to them as well.

**Comment by mcollina** [comment]
Specifically, I would like to provide this escape hatch for:
- https://github.com/fastify/fastify/blob/main/types/reply.d.ts#L76
- https://github.com/fastify/fastify/blob/main/types/instance.d.ts#L120 (this doesn't even have a `.then()` to avoid people reporting this in issues.

The framework is actually designed to handle those correctly as it supports a dual promise/callback API.

**Comment by bradzacher** [comment]
TBH I don't think there's any reason the rule needs to specifically look for a "thenable" - it could specifically look for a type that extends `Promise` (or that's just named promise?)

I don't remember the context (from 5 years ago) for why we made it check for a thenable rather than just promises. The TSLint rule it was derived from did promises by name (with a configurable list) so 🤷.
Would be a breaking change but we could target it for the next major and it would remove false positives from "promise likes".

**Comment by mcollina** [comment]
For some context on the Fastify issue: https://github.com/typescript-eslint/typescript-eslint/issues/2640. Imho adding some kind of branding that allows to communicate to typescript-eslint to treat these as optional would be brilliant.

**Comment by arka1002** [comment]
Hi :wave:, I just had 2 questions.


1. Will these be the preferred reports by the rule ? I'm asking this because I'm very unfamiliar with redux  
The valids -  

```ts
// 1.
type Foo = Promise<number> & { hey?: string }; // it is exported from index.d.ts file of `foo-lib` npm package
/* eslint "@typescript-eslint/no-floating-promises": ["error", { "allowForKnownSafePromises": [ { "from": "package", "name": "Foo", "package": "foo-lib" } ] } ] */
let buzz: Foo = Promise.resolve(5);
buzz; // no error (?)

// 2.
interface Bar extends Promise<number> { hey?: string }; // it is exported from index.d.ts file of `bar-lib` npm package
/* eslint "@typescript-eslint/no-floating-promises": ["error", { "allowForKnownSafePromises": [ { "from": "package", "name": "Bar", "package": "bar-lib" } ] } ] */
let guzz: Bar = Promise.resolve(5);
guzz; // no error (?)

// 3.
type SafeForExceptions = Promise<number>; // it is exported from index.d.ts file of `buzz-lib` npm package
/* eslint "@typescript-eslint/no-floating-promises": ["error", { "allowForKnownSafePromises": [ { "from": "package", "name": "SafeForExceptions", "package": "buzz-lib" } ] } ] */
let shrof: SafeForExceptions = Promise.resolve(5);
shrof; // no error (?)
```

The invalids will be as usual ?

---

2. Is the [old rfc](https://github.com/typescript-eslint/typescript-eslint/discussions/6017#discussion-4576461) open just to include Node.js builtin modules in the *Package* option? Currently [`prefer-readonly-parameter-types`](https://typescript-eslint.io/rules/prefer-readonly-parameter-types/) doesnt support Node.js builtin modules([stackblitz link](https://stackblitz.com/edit/stackblitz-starters-tnrbtd?file=src%2Findex.ts), run `npx eslint src/index.ts` in the terminal, goto that file, see line 4).  
This is referenced from StyleShit's comment - https://github.com/typescript-eslint/typescript-eslint/pull/8502#discussion_r1494653089  
I think (since most of the ts apis are undocumented), currently the *Package* option only handles bare specifiers, can it be open to handle just one absolute specifier, ie `node:` urls ?
